File: 6.zigzag-conversion/solution.java
class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1)
            return s;
        int len = s.length();
        StringBuffer sb = new StringBuffer();
        for (int i = 0; i < numRows; i++)
        {
            if (i >= len)
                break;
        	sb.append(s.charAt(i));
        	for (int j = 1;; j++)
        	{
        		if (i != 0 && i != numRows - 1)
        		{
        			int index = j * (numRows - 1) * 2 - i;
        			if (index >= len)
        				break;
        			sb.append(s.charAt(index));
        		}
        		int index = i + j * (numRows - 1) * 2;
        		if (index >= len)
        			break;
        		sb.append(s.charAt(index));
        	}
        }
        return sb.toString();
    }
}

File: 6.zigzag-conversion/solution.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        result = s[0::(numRows-1)*2]
        max_len = len(s)
        for i in range(1,numRows-1):
            j1 = i
            j2 = (numRows-1)*2 - i
            while True:
                if j1 >= max_len:
                    break
                result += s[j1]
                if j2 >= max_len:
                    break
                result += s[j2]
                j1 += (numRows-1)*2
                j2 += (numRows-1)*2
        result += s[(numRows-1)::(numRows-1)*2]
        return result
